### PR TITLE
Custom continuous query options per query rather than per node

### PR DESCRIPTION
**Pull request by jsternberg**
#5136

### CODE REVIEW

File: influxql/parser.go [hunk 1]
```diff
@@ -1403,6 +1403,13 @@ func (p *Parser) parseCreateContinuousQueryStatement() (*CreateContinuousQuerySt
 	}
 	stmt.Database = ident
 
+	if p.parseTokenMaybe(RESAMPLE) {
```
Comment (gunnaraasen) [inline influxql/parser.go]: Why `parseTokenMaybe` here? Instead of

```
tok, _, _ := p.scanIgnoreWhitespace()
if tok == RESAMPLE {
    stmt.ResampleEvery, stmt.ResampleFor, err = p.parseResample()
} else {
    p.unscan()
}
```

File: services/continuous_querier/config.go [hunk 2]
```diff
@@ -22,42 +19,20 @@ type Config struct {
...
-	// group by times will be computed. For example, if you have RunsPerInterval set to 10 and this setting
-	// to 1m. Then for a group by time(1m) will actually only get computed once per interval (and once per PreviousN).
-	// If you have a group by time(5m) then you'll get five computes per interval. Any group by time window larger
-	// than 10m will get computed 10 times for each interval.
-	ComputeNoMoreThan toml.Duration `toml:"compute-no-more-than"`
+	// Run interval for checking continuous queries. This should be set to the least common factor
+	// of the interval for running continuous queries. If you only aggregate continuous queries
+	// every minute, this should be set to 1 minute. The default is set to '1s' so the interval
+	// is compatible with most aggregations.
+	// This option will be removed in the future after the continuous query engine is refactored
```
Comment (gunnaraasen) [inline services/continuous_querier/config.go]: No point mentioning this right now.

File: influxql/parser.go (hunk 1 above)
Comment (jsternberg) [inline influxql/parser.go]: I noticed it was a very common pattern in the parser to do your suggestion. While that method works, I was hoping to introduce a new convenience function that does the same thing and reduce some of the bloat and variable shadowing the current method introduces.

File: services/continuous_querier/config.go (hunk 2 above)
Comment (jsternberg) [inline services/continuous_querier/config.go]: Removed the last line from the comment.

File: services/continuous_querier/service.go [hunk 3]
```diff
@@ -167,8 +167,10 @@ func (s *Service) Run(database, name string, t time.Time) error {
 		// Loop through CQs in each DB executing the ones that match name.
 		for _, cq := range db.ContinuousQueries {
 			if name == "" || cq.Name == name {
-				// Reset the last run time for the CQ.
-				s.lastRuns[cq.Name] = time.Time{}
+				// Remove the last run time for the CQ
+				if _, ok := s.lastRuns[cq.Name]; ok {
+					delete(s.lastRuns, cq.Name)
```
Comment (gunnaraasen) [inline services/continuous_querier/service.go]: Why this change?

Comment (jsternberg) [inline services/continuous_querier/service.go]: The initial run of a continuous query acts differently than one that has previous been run. One that has previously been run will try to calculate the optimal time it _should_ have run the next query (so as to calculate the oldest interval to regenerate). The initial run ignores this.

This line of code set the time to January 1, 1970 and caused the code attempting to calculate the intervals to try running every query from 1970. Instead of trying to differentiate a default value vs a real value, I just decided removing it from the map was a better solution.

**Comment by jsternberg** [comment]
I think this is ready for review and (potential) merging. I have updated the changelog and the documentation in INFLUXQL.md.

@gunnaraasen @dgnorton @pauldix

File: influxql/parser.go [hunk 4]
```diff
@@ -2391,6 +2398,47 @@ func (p *Parser) parseCall(name string) (*Call, error) {
...
+	}
+
+	// Neither EVERY or FOR were read, so read the next token again
+	// so we can return a suitable error message.
+	if resampleInterval == 0 && resampleMaxDuration == 0 {
+		tok, pos, lit := p.scanIgnoreWhitespace()
+		return 0, 0, newParseError(tokstr(tok, lit), []string{"EVERY", "FOR"}, pos)
+	}
+	return resampleInterval, resampleMaxDuration, nil
+}
```
Comment (benbjohnson) [inline influxql/parser.go]: Prefixing variables with `resample` in this function seems redundant since it's in the `parseResample()` function.

Comment (jsternberg) [inline influxql/parser.go]: Removed the `resample` prefix.

File: services/continuous_querier/service.go [hunk 5]
```diff
@@ -268,46 +264,45 @@ func (s *Service) ExecuteContinuousQuery(dbi *meta.DatabaseInfo, cqi *meta.Conti
...
 
-	recomputeNoOlderThan := time.Duration(s.Config.RecomputeNoOlderThan)
-
-	for i := 0; i < s.Config.RecomputePreviousN; i++ {
-		// if we're already more time past the previous window than we're going to look back, stop
-		if now.Sub(startTime) > recomputeNoOlderThan {
-			return nil
+	// Calculate and set the time range for the query. Go from most recent to least.
+	startTime := now.Add(-resampleEvery).Truncate(interval)
+	for ; startTime.UnixNano() >= oldestTime.UnixNano(); startTime = startTime.Add(-interval) {
```
Comment (benbjohnson) [inline services/continuous_querier/service.go]: What about doing `oldestTime.Before(startTime)` for the `for` condition?

File: services/continuous_querier/service.go [hunk 6]
```diff
@@ -342,13 +337,20 @@ func (s *Service) runContinuousQueryAndWriteResult(cq *ContinuousQuery) error {
...
 	q        *influxql.SelectStatement
 }
 
 func (cq *ContinuousQuery) intoRP() string      { return cq.q.Target.Measurement.RetentionPolicy }
 func (cq *ContinuousQuery) setIntoRP(rp string) { cq.q.Target.Measurement.RetentionPolicy = rp }
 
+type ResampleOptions struct {
+	Every time.Duration
+	For   time.Duration
+}
```
Comment (benbjohnson) [inline services/continuous_querier/service.go]: Can you add a brief comment about the resampling options?

File: services/continuous_querier/service_test.go [hunk 7]
```diff
@@ -92,6 +93,84 @@ func TestContinuousQueryService_Run(t *testing.T) {
...
+	} else if cq.Resample.For != 2*time.Minute {
+		t.Errorf("expected resample for 2m, got %s", influxql.FormatDuration(cq.Resample.For))
+	}
+
+	// Set RunInterval high so we can trigger using Run method.
+	s.RunInterval = 10 * time.Minute
+
+	done := make(chan struct{})
+	expectCallCnt := 0
+	callCnt := 0
```
Comment (benbjohnson) [inline services/continuous_querier/service_test.go]: Minor nit: we typically use `n` to denote count. e.g. `expectCallN`, `callN`.

File: services/continuous_querier/service.go (hunk 5 above)
Comment (jsternberg) [inline services/continuous_querier/service.go]: Sounds good to me. I'll update the PR.

**Comment by benbjohnson** [comment]
Overall, lgtm. :+1:

File: services/continuous_querier/service_test.go (hunk 7 above)
Comment (jsternberg) [inline services/continuous_querier/service_test.go]: I copied this pattern from earlier in the file, so it might be a good idea to go over this file again at some point and get this file using the same conventions as the rest of the codebase.

File: services/continuous_querier/service.go (hunk 6 above)
Comment (jsternberg) [inline services/continuous_querier/service.go]: Done.

File: influxql/parser.go (hunk 1 above)
Comment (dgnorton) [inline influxql/parser.go]: Like the idea.  I think `parseOptionalToken` is a better name.

File: services/continuous_querier/service.go (hunk 5 above)
Comment (jsternberg) [inline services/continuous_querier/service.go]: Done. It needs to be `!startTime.Before(oldestTime)`, but it works.

File: influxql/INFLUXQL.md [hunk 8]
```diff
@@ -229,6 +229,7 @@ ALTER RETENTION POLICY policy1 ON somedb DURATION 1h REPLICATION 4
 
 ```
 create_continuous_query_stmt = "CREATE CONTINUOUS QUERY" query_name on_clause
+                               [ "RESAMPLE" [ "EVERY" duration_lit ] [ "FOR" duration_lit ] ]
```
Comment (dgnorton) [inline influxql/INFLUXQL.md]: Need to fix this.  It allows `CREATE CONTINUOUS QUERY foo ON bar RESAMPLE BEGIN...`

File: influxql/parser_test.go [hunk 9]
```diff
@@ -1051,7 +1051,7 @@ func TestParser_ParseStatement(t *testing.T) {
 
 		// CREATE CONTINUOUS QUERY ... INTO <measurement>
 		{
-			s: `CREATE CONTINUOUS QUERY myquery ON testdb BEGIN SELECT count(field1) INTO measure1 FROM myseries GROUP BY time(5m) END`,
+			s: `CREATE CONTINUOUS QUERY myquery ON testdb RESAMPLE EVERY 1m FOR 1h BEGIN SELECT count(field1) INTO measure1 FROM myseries GROUP BY time(5m) END`,
```
Comment (dgnorton) [inline influxql/parser_test.go]: Should also add tests for just having an `EVERY` clause and just a `FOR` clause.

File: influxql/INFLUXQL.md (hunk 8 above)
Comment (jsternberg) [inline influxql/INFLUXQL.md]: I attempted to make this more clear, but it's very difficult to do a "at least one, but in this order" in BNF since that's not really a thing in context free grammars. It's possible in a context-free grammar, but it looks weird since you basically have to say `S = EF | E | F`.

File: influxql/parser_test.go (hunk 9 above)
Comment (jsternberg) [inline influxql/parser_test.go]: Added two more tests that test each possibility of this.

**Comment by dgnorton** [comment]
+1
